storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,181 @@
+        List<GrasSimulatedUser> users = new ArrayList<>();
+
+        for (User u : userMap.values()) {
+            users.add(getSimulatedUserFor(u, cluster));
+        }
+
+        while (!users.isEmpty()) {
+            Collections.sort(users, new GrasSimulatedUserComparator(cpuAvail, memAvail, genericAvail));
+            GrasSimulatedUser u = users.get(0);
+            TopologyDetails td = u.getNextHighest();
+            if (td == null) {
+                users.remove(0);
+            } else {
+                double score = u.getScore(cpuAvail, memAvail, genericAvail);
+                td = u.simScheduleNextHighest();
+                LOG.info("GRAS SIM Scheduling {} with score of {}", td.getId(), score);
+                cpuAvail -= td.getTotalRequestedCpu();
+                memAvail -= (td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap());
+                for (Map.Entry<String, Double> entry : td.getTotalRequestedGenericResources().entrySet()) {
+                    String resource = entry.getKey();
+                    Double requestedAmount = entry.getValue();
+                    if (!genericAvail.containsKey(resource)) {
+                        LOG.warn("Resource: {} is not supported in this cluster. Ignoring this request.", resource);
+                    } else {
+                        genericAvail.put(resource, genericAvail.get(resource) - requestedAmount);
+                    }
+                }
+                allUserTopologies.add(td);
+            }
+        }
+        return allUserTopologies;
+    }
+
+
+
+    protected static class GrasSimulatedUser extends SimulatedUser {
+
+        // Extend support for Generic Resources in addition to CPU and Memory
+        public final Map<String, Double> guaranteedGenericResources;                // resource name : guaranteed amount
+        private Map<String, Double> assignedGenericResources = new HashMap<>();     // resource name : assigned amount
+
+        public GrasSimulatedUser(User other, ISchedulingState cluster) {
+            super(other, cluster);
+
+            Map<String, Double> guaranteedGenericResources = new HashMap<>();
+            Map<String, Double> availGenericResources = cluster.getClusterTotalGenericResource();   // generic resources that are offered
+            for (Map.Entry<String, Double> entry : availGenericResources.entrySet()) {
+                String resource = entry.getKey();
+                Double guaranteedAmount = other.getGenericGuaranteed().getOrDefault(resource, 0.0);
+                guaranteedGenericResources.put(resource, guaranteedAmount);
+            }
+            this.guaranteedGenericResources = guaranteedGenericResources;
+        }
+
+        @Override
+        public TopologyDetails simScheduleNextHighest() {
+            TopologyDetails td = tds.pop();
+            assignedCpu += td.getTotalRequestedCpu();
+            assignedMemory += td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap();
```

Comment:
We could reply on `super.simScheduleNextHighest();` so everything in GrasSimulatedUser is gras specific, like what you did elsewhere